lib/TTT/validation.ex

```diff
@@ -39,6 +39,23 @@ defmodule Validation do
     end
   end
 
+  def choose_first_player(io, player_token, opponent_token) do
+    io.retrieve(
+      Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
+    )
+    |> clean()
+    |> validate_first_player(io, player_token, opponent_token)
```

Comment:
Would it make more sense to throw an error in validate, then have this case off of it and recurse within itself?

Reply:
like this?

``` 
  def choose_first_player(io, player_token, opponent_token) do
    first_player =
      io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()

    case validate_first_player(first_player, player_token, opponent_token) do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```

Reply:
Yeah, or piping validate_first_player to `|> clean()`

```
def choose_first_player(io, player_token, opponent_token) do
    io.retrieve(
        Messages.get_message(:choose_first_player) <> "(#{player_token} or #{opponent_token}): "
      )
      |> clean()
      |> validate_first_player(player_token, opponent_token)
      |> case do
      {:ok} ->
        first_player

      {:error} ->
        io.display(Messages.get_message(:first_player_error))
        choose_first_player(io, player_token, opponent_token)
    end
  end

  def validate_first_player(first_player, player_token, opponent_token) do
    case first_player do
      ^player_token ->
        {:ok}

      ^opponent_token ->
        {:ok}

      _ ->
        {:error}
    end
  end
```